File: src/Server/Size.cs
namespace Belin.UI;

using System.Text.Json.Serialization;

/// <summary>
/// Defines the size of an element.
/// </summary>
[JsonConverter(typeof(JsonStringEnumConverter))]
public enum Size {

	/// <summary>
	/// An extra small size.
	/// </summary>
	ExtraSmall,

	/// <summary>
	/// A small size.
	/// </summary>
	Small,

	/// <summary>
	/// A medium size.
	/// </summary>
	Medium,

	/// <summary>
	/// A large size.
	/// </summary>
	Large,

	/// <summary>
	/// An extra large size.
	/// </summary>
	ExtraLarge,

	/// <summary>
	/// An extra extra large size.
	/// </summary>
	ExtraExtraLarge
}

/// <summary>
/// Provides extension members for element sizes.
/// </summary>
public static class SizeExtensions {
	extension(Size size) {

		/// <summary>
		/// The CSS class of this size.
		/// </summary>
		public string CssClass => size switch {
			Size.ExtraSmall => "xs",
			Size.Small => "sm",
			Size.Medium => "md",
			Size.Large => "lg",
			Size.ExtraLarge => "xl",
			Size.ExtraExtraLarge => "xxl"
		};
	}
}

File: src/Client/Size.ts
/**
 * Defines the size of an element.
 */
export const Size = Object.freeze({

	/**
	 * An extra small size.
	 */
	ExtraSmall: "ExtraSmall",

	/**
	 * A small size.
	 */
	Small: "Small",

	/**
	 * A medium size.
	 */
	Medium: "Medium",

	/**
	 * A large size.
	 */
	Large: "Large",

	/**
	 * An extra large size.
	 */
	ExtraLarge: "ExtraLarge",

	/**
	 * An extra extra large size.
	 */
	ExtraExtraLarge: "ExtraExtraLarge"
});

/**
 * Defines the size of an element.
 */
export type Size = typeof Size[keyof typeof Size];

/**
 * Returns the CSS class of the specified size.
 * @param size The size.
 * @returns The CSS class of the specified size.
 */
export function cssClass(size: Size): string {
	switch (size) {
		case Size.ExtraSmall: return "xs";
		case Size.Small: return "sm";
		case Size.Large: return "lg";
		case Size.ExtraLarge: return "xl";
		case Size.ExtraExtraLarge: return "xxl";
		default: return "md";
	}
}
